### PR TITLE
Let Vec be used as an argument in fluent strings

### PR DESCRIPTION
**Pull request by mejrs**
An alternative approach would be to have some kind of `#[list]` attribute for the derive macro to use, or a newtype for doing this. I chose the current approach because it's what people will try first and it'd be nice to have it "Just Work" - there's no need to write or read documentation.

### CODE REVIEW

File: tests/ui/issues/issue-15381.stderr [hunk 1]
```diff
@@ -2,7 +2,7 @@ error[E0005]: refutable pattern in `for` loop binding
   --> $DIR/issue-15381.rs:4:9
    |
 LL |     for &[x,y,z] in values.chunks(3).filter(|&xs| xs.len() == 3) {
-   |         ^^^^^^^^ patterns `&[]`, `&[_]`, `&[_, _]` and 1 more not covered
+   |         ^^^^^^^^ patterns `&[]`, `&[_]`, `&[_, _]`, and `&[_, _, _, _, ..]` not covered
```
Comment (mejrs) [inline tests/ui/issues/issue-15381.stderr:5]: I have no idea where this comma in `, and` comes from 🤔

Comment (scottmcm) [reply tests/ui/issues/issue-15381.stderr:5]: Oxford 😏

**Comment by Nilstrieb** [comment]
I think @Manishearth wanted to use ICU4X for this IIRC

**Comment by mejrs** [comment]
> I think @Manishearth wanted to use ICU4X for this IIRC

This uses the list support added in https://github.com/rust-lang/rust/pull/104047 . Other than the "x more" part, which I couldn't figure out - is that even supported?

**Review by Manishearth — CHANGES_REQUESTED** [review]
This shouldn't be done this way: to handle "or more" you should be having the logic in the fluent file, something like


```ftl
mir_build_uncovered = {$truncated ->
        [true] {$witnesses}
        *[other] {$witnesses}, and more
    } not covered
```

and the truncation logic being handled in Rust. Except that's not so straightforward since the list will already format with a `, and`. You _might_ need to define a fluent custom function that is able to do list formatting, cc @davidtwco. Unfortunatley it doesn't have a list type yet.

Furthermore, not all lists will want to be truncated. We should have a `TruncatedDiagnosticList` type that has all this logic.

This would be easier if Fluent had builtin list formatting support (https://github.com/projectfluent/fluent/issues/79)


------


I think the way to do this _for now_ is have another internal list type (perhaps we can merge it all into StrList, which has And/Or and truncate/don't truncate options), and have it pull from a `list_more = and {$count} more` Fluent key that's "global"

**Comment by mejrs** [comment]
Thanks, I've just created a new type for it.

> and have it pull from a `list_more = and {$count} more` Fluent key that's "global"

Where should this live?

**Comment by davidtwco** [comment]
> Where should this live?

We could create a `common.ftl` for it?

**Comment by mejrs** [comment]
I'm not sure how to make it fit together though. I assume I'd need to get access to the bundle inside the `IntoDiagnosticArg ` impl, but that doesn't seem quite right. Is my approach wrong?

**Comment by davidtwco** [comment]
> I'm not sure how to make it fit together though. I assume I'd need to get access to the bundle inside the `IntoDiagnosticArg ` impl, but that doesn't seem quite right. Is my approach wrong?

Is it possible with the way we use `IntoDiagnosticArg` that you could give it a reference to the `Handler` and use `eager_translate`?

**Pull request closed without merging**